src/robots/mecanum_pca_9685.cc

```diff
@@ -0,0 +1,85 @@
+#ifdef __linux__
+// BoB robotics includes
+#include "robots/mecanum_pca_9685.h"
+
+// Standard C includes
+#include <cmath>
+#include <cstdint>
+
+// Standard C++ includes
+#include <algorithm>
+#include <vector>
+
+
+namespace BoBRobotics {
+namespace Robots {
+const PCA9685::Channel MecanumPCA9685::motorChannels[MOTOR_MAX][MOTOR_CHANNEL_MAX] = {
+        {PCA9685::Channel::CHANNEL_0, PCA9685::Channel::CHANNEL_1},
+        {PCA9685::Channel::CHANNEL_2, PCA9685::Channel::CHANNEL_3},
+        {PCA9685::Channel::CHANNEL_4, PCA9685::Channel::CHANNEL_5},
+        {PCA9685::Channel::CHANNEL_6, PCA9685::Channel::CHANNEL_7}};
+//----------------------------------------------------------------------------
+MecanumPCA9685::MecanumPCA9685(int slaveAddress, const char *path)
+  : m_PCA9685(slaveAddress, path)
+{
+}
+//----------------------------------------------------------------------------
+MecanumPCA9685::~MecanumPCA9685()
+{
+    stopReadingFromNetwork();
+    stopMoving();
+}
+//----------------------------------------------------------------------------
+// Omni2D virtuals
+//----------------------------------------------------------------------------
+void
+MecanumPCA9685::omni2D(float forward, float sideways, float turn)
+{
+    // Cache left and right
+    setWheelSpeed(forward, sideways, turn);
+
+    // resolve to motor speeds
+    const float m1 = -sideways + forward - turn;
+    const float m2 = +sideways + forward + turn;
+    const float m3 = +sideways + forward - turn;
+    const float m4 = -sideways + forward + turn;
+
+    driveMotors(m1, m2, m3, m4);
+}
+//----------------------------------------------------------------------------
+void MecanumPCA9685::driveMotors(float m1, float m2, float m3, float m4)
+{
+    // clamp values to be between -1 and 1 after resolving
+    const auto cap = [](float &val) { val = std::min(1.f, std::max(val, -1.f)); };
+    cap(m1);
+    cap(m2);
+    cap(m3);
+    cap(m4);
+
+    setMotorThrottle(Motor::MOTOR_1, m1);
+    setMotorThrottle(Motor::MOTOR_2, m2);
+    setMotorThrottle(Motor::MOTOR_3, m3);
+    setMotorThrottle(Motor::MOTOR_4, m4);
+}
+//----------------------------------------------------------------------------
+void MecanumPCA9685::setMotorThrottle(Motor motor, float throttle)
```

Comment:
The problem with ``Motor`` being an ``enum`` rather than an ``enum class`` is that a user could accidentally pass e.g. a value of ``5`` for ``motor`` and the compiler would happily accept it. (You could also just check for this case with ``BOB_ASSERT()``).

Reply:
True - as it's an old-style enum I should check - bleurgh